aggregator/follower_flush_mgr.go

```diff
@@ -0,0 +1,350 @@
+		forcedFlushWindowSize: opts.ForcedFlushWindowSize(),
+		logger:                instrumentOpts.Logger(),
+		scope:                 scope,
+		doneCh:                doneCh,
+		proto:                 &schema.ShardSetFlushTimes{},
+		flushTimesState:       flushTimesUninitialized,
+		lastFlushed:           nowFn(),
+		sleepFn:               time.Sleep,
+		metrics:               newFollowerFlushManagerMetrics(scope),
+	}
+	mgr.flushTask = &followerFlushTask{mgr: mgr}
+	return mgr
+}
+
+func (mgr *followerFlushManager) Open(shardSetID string) {
+	mgr.Lock()
+	defer mgr.Unlock()
+
+	mgr.flushTimesKey = fmt.Sprintf(mgr.flushTimesKeyFmt, shardSetID)
+	mgr.openedAt = mgr.nowFn()
+	go mgr.watchFlushTimes()
+}
+
+// NB(xichen): no actions needed for initializing the follower flush manager.
+func (mgr *followerFlushManager) Init([]*flushBucket) {}
+
+func (mgr *followerFlushManager) Prepare(buckets []*flushBucket) (flushTask, time.Duration) {
+	var (
+		shouldFlush        = false
+		flushersByInterval []flushersGroup
+	)
+	// NB(xichen): a flush is triggered in the following scenarios:
+	// * The flush times persisted in kv have been updated since last flush.
+	// * Sufficient time (a.k.a. maxNoFlushDuration) has elapsed since last flush.
+	now := mgr.nowFn()
+	mgr.Lock()
+	if mgr.flushTimesState == flushTimesUpdated {
+		mgr.lastFlushed = now
+		mgr.flushTimesState = flushTimesProcessed
+		flushersByInterval = mgr.flushersFromKVUpdateWithLock(buckets)
+		mgr.Unlock()
+		shouldFlush = true
+		mgr.metrics.kvUpdateFlush.Inc(1)
+	} else {
+		durationSinceLastFlush := now.Sub(mgr.lastFlushed)
+		if durationSinceLastFlush > mgr.maxNoFlushDuration {
+			mgr.lastFlushed = now
+			mgr.Unlock()
+			shouldFlush = true
+			mgr.metrics.forcedFlush.Inc(1)
+			flushBeforeNanos := now.Add(-mgr.maxNoFlushDuration).Add(mgr.forcedFlushWindowSize).UnixNano()
+			flushersByInterval = mgr.flushersFromForcedFlush(buckets, flushBeforeNanos)
+		} else {
+			mgr.Unlock()
+		}
+	}
+
+	if !shouldFlush {
+		return nil, mgr.checkEvery
+	}
+	mgr.flushTask.flushersByInterval = flushersByInterval
+	return mgr.flushTask, 0
+}
+
+// NB(xichen): the follower flush manager flushes data based on the flush times
+// stored in kv and does not need to take extra actions when a new bucket is added.
+func (mgr *followerFlushManager) OnBucketAdded(int, *flushBucket) {}
+
+// The follower flush manager may only lead if and only if all the following conditions
+// are met:
+// * The flush times persisted in kv have been read at least once.
+// * There are no outstanding/unprocessed kv flush times update.
+// * All the aggregation windows since the flush manager is opened have ended.
+func (mgr *followerFlushManager) CanLead() bool {
+	mgr.RLock()
+	defer mgr.RUnlock()
+
+	if mgr.flushTimesState != flushTimesProcessed {
+		mgr.metrics.flushTimesNotProcessed.Inc(1)
+		return false
+	}
+
+	allWindowsEnded := true
+	for _, shardFlushTimes := range mgr.proto.ByShard {
```

Comment:
Is this safe around a topology change? For example, when another shardSet was removed from the placement, this shardset may get assigned a few more shards, what if the follower can lead the old shards but not the new ones?